### PR TITLE
Bug 1241194 - Better presentation for the first time in the task{-graph} inspector

### PR DESCRIPTION
**Pull request by kristelteng**
@walac @djmitche : 
This is the new UI for task{-graph} inspector. It allows the users to see the previous task that they have  visited. Instead of using cookie, I have used localstorage. 

For now, I have created 2 components for task inspector and task-graph inspector. I still to refactor the code, to make use 1 component only.

### CODE REVIEW

File: lib/ui/previoustaskgraph.jsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        items.splice(prevIndex, 1);
+      }
+      items.push(newTaskId);
+
+      while (items.length > 5) {
+        items.shift();
+      } 
+
+      localStorage.setItem('graph-items', JSON.stringify(items));
+      }
```
Comment (walac) [inline lib/ui/previoustaskgraph.jsx]: nit: indentation

File: lib/ui/previoustaskgraph.jsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+      return items;
+    },
```
Comment (walac) [inline lib/ui/previoustaskgraph.jsx]: nit: indentation

File: lib/ui/previoustaskgraph.jsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    };
+  },
+
+  render() {
+    var list = [];
+    var taskGraphIds = this.state.previousTaskIds || [];
+    taskGraphIds.forEach(taskGraphId => {
+      var link = "#" + taskGraphId;
+      list.push(<li key={taskGraphId}><a href={link}>{taskGraphId}</a></li>);
+      });
```
Comment (walac) [inline lib/ui/previoustaskgraph.jsx]: nit: in taskcluster, our convention is to indent closing braces/parenthesis at the same level as the calling function.

File: lib/ui/previoustaskgraph.jsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    list.reverse();
+
+    if (list.length === 0) {
+      return (
+        <span>
+          <p>Find tasks at <a href="/index">index</a></p>
```
Comment (walac) [inline lib/ui/previoustaskgraph.jsx]: hrm, actually `index` does not provide links to task graphs, so this should be replaced by something else. @djmitche, suggestions?

File: lib/ui/previoustasks.jsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  mixins: [
+    // Calls load() initially and on reload()
+    utils.createTaskClusterMixin({
+      // Reload when props.status.taskId changes, ignore credential changes
+      reloadOnProps:           ['taskId'],
+      reloadOnLogin:          false
+    })
+  ],
+
+  getPreviousTaskIds(newTaskId) {
```
Comment (walac) [inline lib/ui/previoustasks.jsx]: This function is exactly the same at the previoustaskgraph module, except by the key used in `localStorage`. Better putting it in a common module and reuse it.

File: task-graph-inspector/taskgraphinspector.jsx [hunk 6]
```diff
@@ -38,17 +39,17 @@ var TaskGraphInspector = React.createClass({
       keys:                       ['taskGraphId', 'taskId'],
       type:                       'string'
     })
-  ],
+    ],
```
Comment (walac) [inline task-graph-inspector/taskgraphinspector.jsx]: Why this?

File: task-graph-inspector/taskgraphinspector.jsx [hunk 7]
```diff
@@ -38,17 +39,17 @@ var TaskGraphInspector = React.createClass({
...
-      taskGraphIdInput:   ''
+    /** Get initial state */
+    getInitialState: function() {
+      return {
+        taskGraphId:        '',
+        taskId:             null,
+        taskGraphLoaded:    true,
+        taskGraphError:     undefined,
+        taskGraph:          null,
+        taskGraphIdInput:   ''
```
Comment (walac) [inline task-graph-inspector/taskgraphinspector.jsx]: Any reason to realign this?

File: task-graph-inspector/taskgraphinspector.jsx [hunk 8]
```diff
@@ -82,7 +83,7 @@ var TaskGraphInspector = React.createClass({
     var promisedState = {};
     var key = 'task/' + this.state.taskId + '/status';
     promisedState[key] = this.queue.status(this.state.taskId)
-                                   .then(_.property('status'));
+    .then(_.property('status'));
```
Comment (walac) [inline task-graph-inspector/taskgraphinspector.jsx]: Same here

File: task-graph-inspector/taskgraphinspector.jsx [hunk 9]
```diff
@@ -107,17 +108,17 @@ var TaskGraphInspector = React.createClass({
...
+    this.queueEvents.taskPending(qkey),
+    this.queueEvents.taskRunning(qkey),
+    this.queueEvents.artifactCreated(qkey),
+    this.queueEvents.taskCompleted(qkey),
+    this.queueEvents.taskFailed(qkey),
+    this.queueEvents.taskException(qkey),
+    this.schedulerEvents.taskGraphRunning(skey),
+    this.schedulerEvents.taskGraphExtended(skey),
+    this.schedulerEvents.taskGraphBlocked(skey),
+    this.schedulerEvents.taskGraphFinished(skey)
```
Comment (walac) [inline task-graph-inspector/taskgraphinspector.jsx]: Why was the indentation removed?

File: task-graph-inspector/taskgraphinspector.jsx [hunk 10]
```diff
@@ -125,13 +126,13 @@ var TaskGraphInspector = React.createClass({
...
-      this.queueEvents.taskCompleted().exchange,
-      this.queueEvents.taskFailed().exchange,
-      this.queueEvents.taskException().exchange
+    this.queueEvents.taskDefined().exchange,
+    this.queueEvents.taskPending().exchange,
+    this.queueEvents.taskRunning().exchange,
+    this.queueEvents.artifactCreated().exchange,
+    this.queueEvents.taskCompleted().exchange,
+    this.queueEvents.taskFailed().exchange,
+    this.queueEvents.taskException().exchange
```
Comment (walac) [inline task-graph-inspector/taskgraphinspector.jsx]: More indentation issues :p

File: task-graph-inspector/taskgraphinspector.jsx [hunk 11]
```diff
@@ -457,10 +468,10 @@ var TaskGraphInspector = React.createClass({
...
       <TaskView ref="taskView"
-          hashEntry={this.nextHashEntry()}
-          status={status}
-          queue={this.props.queue}/>
-    );
+      hashEntry={this.nextHashEntry()}
+      status={status}
+      queue={this.props.queue}/>
+      );
   },
```
Comment (walac) [inline task-graph-inspector/taskgraphinspector.jsx]: Summarizing here: the file has a lot of indentation changes for no apparent reason, maybe IDE config?

File: task-graph-inspector/taskgraphinspector.jsx [hunk 12]
```diff
@@ -374,74 +385,74 @@ var TaskGraphInspector = React.createClass({
...
+          relation = <span className={stateLabel}>required</span>;
         }
-        </tbody>
-      </table>
-    );
-  },
-
-  /** Render taskview for currently selected task */
-  renderTaskView: function() {
+        if (dependentTasks.indexOf(task.taskId) !== -1) {
```
Comment (walac) [inline task-graph-inspector/taskgraphinspector.jsx]: Indentation should be two spaces.

**Comment by walac** [comment]
\o/
A bunch of small nit comments, but overall it looks great! @djmitche might have additional comments. Thanks for the patch!

File: lib/ui/previoustaskgraph.jsx (hunk 4 above)
Comment (djmitche) [inline lib/ui/previoustaskgraph.jsx]: Maybe "Find tasks at _index_, then click their "TaskGroupId" link to see task graphs".  It's the best I can think of..

File: lib/ui/previoustasks.jsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+      previousTaskIds: [],
+      taskId: ''
+    };
+  },
+
+  /** Render row with role */
+  render() {
+    var rows = [];
+    var taskIds = this.state.previousTaskIds || [];
+    taskIds.forEach(taskId => {
```
Comment (jonasfj) [inline lib/ui/previoustasks.jsx]: Use `.map` instead, it's much easier to follow...

**Comment by jonasfj** [comment]
@kristelteng indentation is two spaces.. And as wcosta points out this PR needs a bit of cleanup before we can review it..

Also am I the only who can't run this? I get `index.js:6Uncaught TypeError: Cannot read property 'Promise' of undefined` when running task-inspector locally...

**Comment by djmitche** [comment]
@jonasfj that's usually caused by using the wrong npm version.  It works with node 4.2.6 and npm 0.31.0, although you'll probably have to delete node_modules and reinstall.

**Comment by jonasfj** [comment]
ah, could be... retrying with node 4 :)

**Comment by jonasfj** [comment]
works with node 4... my bad...
let's cleanup some of the indentation and other minor issues, then we can land this.

File: lib/ui/previoustaskgraph.jsx (hunk 4 above)
Comment (kristelteng) [inline lib/ui/previoustaskgraph.jsx]: Ok will make the necessary changes

File: lib/ui/previoustasks.jsx (hunk 13 above)
Comment (kristelteng) [inline lib/ui/previoustasks.jsx]: Thanks, will make the necessary changes

**Comment by kristelteng** [comment]
@walac @djmitche: Can you please review the revised PR. Thanks.

**Comment by walac** [comment]
lgtm!

@djmitche @jonasfj if it is ok for you, we can merge it

@kristelteng great work, thanks!

**Comment by djmitche** [comment]
Very nice!

**Comment by kristelteng** [comment]
@walac @djmitche : Great thanks.

**Pull request merged**